lib/mayaUsd/fileio/writeJobContext.cpp

```diff
@@ -476,16 +476,10 @@ UsdMayaWriteJobContext::CreatePrimWriter(
     SdfPath writePath = usdPath;
 
     try {
-        const MFnDagNode& dagNodeFn =
-            dynamic_cast<const MFnDagNode&>(depNodeFn);
+        const MDagPath dagPath =
+            UsdMayaUtil::getDagPath(depNodeFn, /* reportError = */ false);
 
-        MStatus status;
-        const MDagPath dagPath = dagNodeFn.dagPath(&status);
-        if (status != MS::kSuccess || !dagPath.isValid()) {
-            TF_CODING_ERROR(
-                "Invalid MDagPath for MFnDagNode '%s'. Verify that it was "
-                "constructed using an MDagPath.",
-                dagNodeFn.fullPathName().asChar());
+        if (!dagPath.isValid()) {
             return nullptr;
```

Comment:
The `try`/`catch` that wraps this whole thing should go away now, since that's handled by `getDagPath()`. The coding error that's in the catch below should go here.

Reply:
The coding error is still missing.

Reply:
I was wondering where it would go? So the whole catch block should be replace with this? Not sure if I get the logic there.
```
if (writePath.IsEmpty()) {
    TF_CODING_ERROR(
        "No usdPath supplied for DG node '%s'.",
        UsdMayaUtil::GetMayaNodeName(depNodeFn.object()).c_str());
}
```

Reply:
This function should be able to work with either DAG nodes or DG nodes, and the try/catch that was here previously is what was used to determine that. If you passed in an MFnDependencyNode that represents a DG node, you **must** also pass in an SdfPath. With the current changes, this will now **only** work for DAG nodes, which is incorrect.

`UsdMayaUtil::getDagPath()` is now what tells you whether the node is a DAG node or a DG node based on whether the MDagPath you get back is valid or not. So with that, here's what I think the first part of this function should look like:

```C++
    SdfPath writePath = usdPath;

    const MDagPath dagPath =
        UsdMayaUtil::getDagPath(depNodeFn, /* reportError = */ false);
    if (!dagPath.isValid()) {
        // This must be a DG node. usdPath must be supplied for DG nodes.
        if (writePath.IsEmpty()) {
            TF_CODING_ERROR(
                "No usdPath supplied for DG node '%s'.",
                UsdMayaUtil::GetMayaNodeName(depNodeFn.object()).c_str());
            return nullptr;
        }
    } else {
        if (dagPath.length() == 0u) {
            // This is the world root node. It can't have a prim writer.
            return nullptr;
        }

        if (writePath.IsEmpty()) {
            writePath = ConvertDagToUsdPath(dagPath);
        }

        const MFnDagNode dagNodeFn(dagPath);
        const bool instanced = dagNodeFn.isInstanced(/* indirect = */ false);
        if (mArgs.exportInstances && instanced && !forceUninstance) {
            // Deal with instances -- we use a special internal writer for them.
            return std::make_shared<UsdMaya_InstancedNodeWriter>(
                dagNodeFn,
                writePath,
                *this);
        }
    }

    ...
```

Reply:
@mattyjams  Ahhh..... it definitely makes sense. see b7a202b747ac4f45f7b513e7b1428ba8969f0949